go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package mpt
+
+import (
+	"context"
+	"fmt"
+	"github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/Fantom-foundation/Carmen/go/common/interrupt"
+	"github.com/Fantom-foundation/Carmen/go/common/witness"
+	"golang.org/x/exp/maps"
+)
+
+// VerifyProofArchiveTrie verifies the consistency of witness proofs for an archive trie.
+// It reads the trie for each block of the archive and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots of each block.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func VerifyProofArchiveTrie(ctx context.Context, dir string, config MptConfig, from, to int, observer VerificationObserver) (err error) {
+	trie, err := OpenArchiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = trie.Close()
+	}()
+
+	return verifyProofArchiveTrie(ctx, trie, from, to, observer)
+}
+
+// verifyProofArchiveTrie verifies the consistency of witness proofs for an archive trie.
+// It reads the trie for each block of the archive and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots of each block.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func verifyProofArchiveTrie(ctx context.Context, trie *ArchiveTrie, from, to int, observer VerificationObserver) error {
+	observer.StartVerification()
+	blockHeight, empty, err := trie.GetBlockHeight()
+	if err != nil {
+		observer.EndVerification(err)
+		return err
+	}
+	if empty {
+		return nil
+	}
+
+	if from < 0 {
+		from = 0
+	}
+
+	if to <= 0 || to > int(blockHeight) {
+		to = int(blockHeight) + 1
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total %d blocks", to-from))
+	for i := uint64(from); i < uint64(to); i++ {
+		observer.Progress(fmt.Sprintf("Collecting state for block... \t %d", i))
+		// collect all accounts for this block
+		visitor := accountCollectingVisitor{accounts: accountsMap{}, ctx: ctx}
+		if err := trie.VisitTrie(i, &visitor); err != nil {
+			observer.EndVerification(err)
```

Comment:
I can see call to this func in most return cases. Why dont you put it into a `defer`?
If you want to avoid the call when the app has been interrupted, you can check the interrupt again in the defer. 
```
defer func() {
   if !interrupt.IsCancelled(ctx) {
      observer.EndVerification(err)
   }
}
```

Although I dont see the reason why you want to avoid it in this case.

Reply:
it is called in the if conditions where an error has occurred, to propagate this error. 

I could define an error variable somewhere on a method global level, assign it everywhere when there is an error, and use `defer`, but I do not see it very much readable. 

I think now the error instance is encapsulated in enclosing `if` statement, which I think is more readable.

Reply:
I like this better but I will leave it up to your preference then.